installinator/src/block_size_writer.rs

```diff
@@ -0,0 +1,223 @@
+        // We should never buffer up more than `block_size` data.
+        assert!(self.buf.len() <= self.block_size);
+
+        // If we already have exactly `block_size` bytes, begin by flushing.
+        if self.buf.len() == self.block_size {
+            ready!(self.as_mut().flush_buf(cx))?;
+        }
+
+        let me = self.project();
+        if me.buf.is_empty() {
+            // If our buffer is empty, either directly write `block_size` bytes
+            // from `buf` to `inner` (if there's enough data in `buf`) or just
+            // copy it into our buffer.
+            if buf.len() >= *me.block_size {
+                me.inner.poll_write(cx, &buf[..*me.block_size])
+            } else {
+                // `me.buf` is empty and `buf` is strictly less than
+                // `self.block_size`, so just copy it.
+                me.buf.extend_from_slice(buf);
+                Poll::Ready(Ok(buf.len()))
+            }
+        } else {
+            // Our buffer already has data - just copy as much of `buf` as we
+            // can onto the end of it.
+            let n = usize::min(*me.block_size - me.buf.len(), buf.len());
+            me.buf.extend_from_slice(&buf[..n]);
+            Poll::Ready(Ok(n))
+        }
+    }
+
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        ready!(self.as_mut().flush_buf(cx))?;
+        self.get_pin_mut().poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        ready!(self.as_mut().flush_buf(cx))?;
+        self.get_pin_mut().poll_shutdown(cx)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_helpers::with_test_runtime;
+    use anyhow::Result;
+    use test_strategy::proptest;
+    use tokio::io::AsyncWriteExt;
+
+    // Dummy `AsyncWrite` that always accepts all data and keeps track of the
+    // sizes of the buffers it was given to write.
+    #[derive(Debug, Default)]
+    struct InnerWriter {
+        data: Vec<u8>,
+        write_requests: Vec<usize>,
+    }
+
+    impl AsyncWrite for InnerWriter {
+        fn poll_write(
+            mut self: Pin<&mut Self>,
+            _cx: &mut Context<'_>,
+            buf: &[u8],
+        ) -> Poll<Result<usize, io::Error>> {
+            self.data.extend_from_slice(buf);
+            self.write_requests.push(buf.len());
+            Poll::Ready(Ok(buf.len()))
+        }
+
+        fn poll_flush(
+            self: Pin<&mut Self>,
+            _cx: &mut Context<'_>,
+        ) -> Poll<Result<(), io::Error>> {
+            Poll::Ready(Ok(()))
+        }
+
+        fn poll_shutdown(
+            self: Pin<&mut Self>,
+            _cx: &mut Context<'_>,
+        ) -> Poll<Result<(), io::Error>> {
+            Poll::Ready(Ok(()))
+        }
+    }
+
+    #[proptest]
+    fn proptest_block_writer(
```

Comment:
Nice test :)

Reply:
Thanks! I was happy when I replaced `BlockSizeWriter` with tokio's `BufWriter` and saw the test fail immediately (even when constructed with capacity == block_size).